addons/point_of_sale/static/src/app/store/pos_store.js

```diff
@@ -178,14 +176,142 @@ export class PosStore extends Reactive {
+
+        // Custom data
+        this.partner_commercial_fields = this.data.custom.partner_commercial_fields;
+        this.server_version = this.data.custom.server_version;
+        this.base_url = this.data.custom.base_url;
+        this.has_cash_move_perm = this.data.custom.has_cash_move_perm;
+        this.has_available_products = this.data.custom.has_available_products;
+        this.pos_special_products_ids = this.data.custom.pos_special_products_ids;
+        this.open_orders_json = this.data.custom.open_orders;
+        this.models = this.data.models;
+
+        // Add Payment Interface to Payment Method
+        for (const pm of this.models["pos.payment.method"].getAll()) {
+            const PaymentInterface = this.electronic_payment_interfaces[pm.use_payment_terminal];
+
+            if (PaymentInterface) {
+                pm.payment_terminal = new PaymentInterface(this, pm);
+            }
+        }
+
+        // Create printer with hardware proxy, this will override related model data
+        this.unwatched.printers = [];
+        for (const relPrinter of this.models["pos.printer"].getAll()) {
+            const printer = relPrinter.serialize();
+            const HWPrinter = this.create_printer(printer);
+
+            HWPrinter.config = printer;
+            this.unwatched.printers.push(HWPrinter);
+
+            for (const id of printer.product_categories_ids) {
+                this.printers_category_ids_set.add(id);
+            }
+        }
+        this.config.iface_printers = !!this.unwatched.printers.length;
+
+        // Monitor product pricelist
+        this.computeProductPricelistCache();
+    }
+
+    computeProductPricelistCache() {
+        const date = DateTime.now();
+        const products = this.models["product.product"].getAll();
+        const pricelistItems = this.models["product.pricelist.item"].getAll();
+
+        for (const product of products) {
+            const applicableRules = {};
+
+            for (const item of pricelistItems) {
+                if (!applicableRules[item.pricelist_id.id]) {
+                    applicableRules[item.pricelist_id.id] = [];
+                }
+
+                if (!product.isPricelistItemUsable(item, date)) {
+                    continue;
+                }
+
+                if (item.product_id && product.id === item.product_id.id) {
+                    applicableRules[item.pricelist_id.id].push(item);
+                } else if (
+                    item.raw.product_tmpl_id &&
+                    product.raw?.product_tmpl_id === item.raw.product_tmpl_id
+                ) {
+                    applicableRules[item.pricelist_id.id].push(item);
+                } else if (!item.raw.product_tmpl_id && !item.raw.product_id) {
+                    applicableRules[item.pricelist_id.id].push(item);
+                }
+            }
+
+            product.cachedPricelistRules = applicableRules;
```

Comment:
When linking `product.product` with `product.pricelist.item`, don't worry about the ordering.